ampd/src/starknet/events/contract_call.rs

```diff
@@ -0,0 +1,278 @@
+pub struct ContractCallEvent {
+    pub destination_address: String,
+    pub destination_chain: String,
+    pub source_address: String,
+    pub payload_hash: Hash,
+}
+
+/// An error, representing failure to convert/parse a starknet event
+/// to some specific event.
+#[derive(Error, Debug)]
+pub enum ContractCallError {
+    #[error("Invalid ContractCall event: {0}")]
+    InvalidEvent(String),
+    #[error("Cairo short string parse error: {0}")]
+    Cairo(#[from] ParseCairoShortStringError),
+    #[error("FieldElement operation errored with out of range: {0}")]
+    FeltOutOfRange(#[from] ValueOutOfRangeError),
+    #[error("Failed int conversion: {0}")]
+    TryFromConversion(#[from] TryFromIntError),
+    #[error("ByteArray type error: {0}")]
+    ByteArray(#[from] ByteArrayError),
+}
+
+impl TryFrom<starknet_core::types::Event> for ContractCallEvent {
+    type Error = ContractCallError;
+
+    fn try_from(starknet_event: starknet_core::types::Event) -> Result<Self, Self::Error> {
+        if starknet_event.keys.len() != 2 {
+            return Err(ContractCallError::InvalidEvent(
+                "ContractCall should have exactly 2 event keys - event_type and destination_chain"
+                    .to_owned(),
+            ));
+        }
+
+        // first key is always the event type
+        let event_type_felt = starknet_event.keys[0];
+        if !matches!(
+            EventType::parse(event_type_felt),
+            Some(EventType::ContractCall)
+        ) {
+            return Err(ContractCallError::InvalidEvent(
+                "not a ContractCall event".to_owned(),
+            ));
+        }
+
+        // destination_chain is the second key in the event keys list (the first key
+        // defined from the event)
+        //
+        // This field, should not exceed 252 bits (a felt's length)
+        let destination_chain = parse_cairo_short_string(&starknet_event.keys[1])?;
+
+        // source_address represents the original caller of the `call_contract` gateway
+        // method. It is the first field in data, by the order defined in the
+        // event.
+        //
+        // TODO: Not sure if `064x` is the correct formatting. Maybe we should calculate
+        // the pedersen hash of the felt as described here, to get the actual address,
+        // although I'm not sure that we can do it as described here:
+        // https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/contract-address/
+        let source_address = format!("0x{:064x}", starknet_event.data[0]);
+
+        // destination_contract_address (ByteArray) is composed of FieldElements
+        // from the second element to elemet X.
+        let destination_address_chunks_count_felt = starknet_event.data[1];
+        let destination_address_chunks_count_u32 =
+            u32::try_from(destination_address_chunks_count_felt)?;
+        let da_chunks_count = usize::try_from(destination_address_chunks_count_u32)?;
```

Comment:
nitpick: This conversion should never fail, right? We should be able to safely use `from` or even `as` here.

Reply:
u32 could've failed (clippy didn't complain), but I changed to u8 (because it's a single 252 bit felt), which won't fail.
So I applied your suggestion before reading it.